### ISSUE TITLE
[enhancement] support for dynamically set "desired capabilities" and "cookies" 

### ISSUE THREAD

**Issue by r-roos** [lucianopuccio/golem#88]
@lucianopuccio I don't see any option to set an enhancement label for this issue via the regular issue input from github, so that is why I used [enhancement] in the title.

Looking at the great Zalenium grid container (https://zalando.github.io/zalenium/) and some other remote browser environments,  it would be great if Golem would support the following features: 

- Setting up a browser instance with dynamically set "desired capabilities" per test case, enabling:
-- Custom browser versions / platforms for remote browser tests.
-- Naming of recorded test video's within Zalenium in case the testcase name is set via 'desired capabilities'(see referenced website -test name- section).

- Set dynamic cookies during tests. 
-- In case of Zalenium this would enable referenced test steps in the recorded Video's (see referenced website -Referencing test steps in the recorded video- section-)

**Comment by lucianopuccio** [comment]
Thanks for bringing this up. I would need to understand how would you implement this.

Currently, you can define capabilities in the settings file:
settings.json
```
{
"remote_browsers": {
        "chrome_remote_mac": {
            "browserName": "chrome",
            "platform": "macOS 10.12"
        },
        "firefox_remote_windows": {
            "browserName": "firefox",
            "platform": "Windows 10"
        }
    }
}
```

Then you select the drivers when running tests:
```golem run project test -b chrome_remote_mac firefox_remote_windows```

Or from a suite:
suite.py
```
browsers = [
    'chrome_remote_mac',
    'firefox_remote_windows'
]
tests = [...]
```

Doing this, every test will be executed for each browser. We could enable defining capabilities inside a test:
test.py
```
capablities = {
    "browserName": "chrome",
    "platform": "macOS 10.12"
}

def test(data):
    pass
```

But doing this, should the test capabilities override previous defined capabilities or merge with them?

**Comment by r-roos** [comment]
It would be great if setting capabilities inside a test would act as an override + merge from the global set on in settings.json. 

So in the example of setting a custom test name for the video recording, I imagine the following (in ways  largely similar to your example)

Global browser setting:
```
{
"remote_browsers": {
        "chrome_remote_mac": {
            "browserName": "chrome",
            "platform": "macOS 10.12"
        }
    }
}
```

Settings within the test case, containing extra capabilities and an override:
```
capabilities = {
    "browserName": "firefox",
    "name": "Testcase 01 - Label for zalenium"
}
```

Effectively what golem uses after overriding and merging the settings:

```
capabilities = {
    "browserName": "firefox",
    "platform": "macOS 10.12",
    "name": "Testcase 01 - Label for the zalenium video"
}
```

**Comment by r-roos** [comment]
@lucianopuccio I have tried the existing functionality as available in version 0.4.5 available via pypi. 

There seems to be a bug causing golem to not start tests when processing using custom browser settings. 
I managed to isolate the bug to source file (golem/core/settings_manager.py function get_remote_browsers line 168)
This function should return a dictionary with custom remote browsers and its related settings. Changing the code to return a dictionary will fix the issue for running test cases with custom settings, however breaks the browser input suggestions in the run suites GUI (which actually requires the list)
```
def get_remote_browsers(settings):
#    remote_browsers = list(settings['remote_browsers'].keys())
    remote_browsers = dict(settings['remote_browsers'])
    return remote_browsers
```

I see in the current github source you reworked some of this code, I do however suspect the same bug still exists though.

**Comment by lucianopuccio** [comment]
Sorry for the delay. Indeed there was an issue with get_remote_browsers. It should be fixed now, the fix is not yet in pypi though.

Back to the original issue, I added an action to add capabilities from inside a test, check: https://github.com/lucianopuccio/golem/commit/88ac74a35de87cff7aa726fa1f1ef802e18eff22

The implementation would be something like this:
```
def setup(data):
    set_browser_capability('name', 'Testcase 01 - Label for the zalenium video')

def test(data):
    # ...
```

Cookie management remains to be seen but it could also be implemented by actions.

**Comment by r-roos** [comment]
Thanks for the update and code changes! I will give this a try later today and provide feedback.

To double check: The "def setup" is invoked before the browser session is started ?

**Comment by r-roos** [comment]
@lucianopuccio I have tested the new set_browser_capability action and it seems to work like a charm with custom browsers, many thanks for implementing this feature

<img width="1118" alt="screen shot 2018-02-14 at 20 06 48" src="https://user-images.githubusercontent.com/32699953/36225868-61b8515e-11cc-11e8-96cb-3e03fa6dbb02.png">

Question, are you aware on some odd behaviour regarding threading ? (if not I will open a new issue)
Examples:
- Setting up multiple browsers in a test suite will crash the python process with the following message: may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug
- The same error message is observed when: Having tests with multiple data rows like:
<img width="434" alt="screen shot 2018-02-14 at 21 24 00" src="https://user-images.githubusercontent.com/32699953/36226178-6d3acdee-11cd-11e8-81b0-90c52de43a91.png">

**Comment by r-roos** [comment]
@lucianopuccio In the previous post I report forking issues with the current source implementation. It seems these errors are related to the use of macos high sierra and not a bug within golem. Running the latest code on my linux test environment does not show any problems

On macos high sierra the following work around can be used before starting golem
Using the following workaround on [page](https://blog.phusion.nl/2017/10/13/why-ruby-app-servers-break-on-macos-high-sierra-and-what-can-be-done-about-it/) solves the issue.

**Comment by lucianopuccio** [comment]
@r-roos I'm glad that it worked ok. I created an action to add cookies (https://golem-framework.readthedocs.io/en/latest/golem-actions.html#add-cookie-cookie-dict). It uses the selenium API which has some drawbacks. It is available in 0.4.6. Let me know if it solves the original issue.

In regards to the macOS issue, I created #96

**Comment by r-roos** [comment]
@lucianopuccio I ran some tests with the new cookie implementation, looking good! I think we can consider this topic closed and implemented. Many thanks for creating these new actions.

Regarding the macOS issue we can further discuss this in the bug you created.

One last item: You released 0.4.6 today which does not include some fixes which are in the pull request list, which causes breaking issues (at least breaking in my view):
1. Test data from csv files are always single quoted strings (including special types like dict / strings) caused by a missing import
2. Pages which are located in subfolders cannot be imported when running test cases. 

both issues are fixed with PR: https://github.com/lucianopuccio/golem/pull/95

**Comment by r-roos** [comment]
Closing this topic as it is implemented
